### DISCUSSION TITLE
Linux/GTK: DBus Activation & systemd user units

### DISCUSSION THREAD

**Discussion by jcollie** [ghostty-org/ghostty#2442] [Feature Requests, Ideas] [answered]
As an improvement on the GTK single instance support, I'd like to propose adding DBus activation and a systemd user unit. Currently, Ghostty is launched on Linux/GTK using a single `.desktop` file. That `.desktop` file has an `Exec=` entry that tells the desktop manager how to launch Ghostty. Subsequent launches using the `.desktop` file rely on Ghostty detecting that an existing instance of Ghostty is already running and forwarding a new window request to the existing Ghostty instance.

DBus activation changes that. A `.desktop` file is marked as DBus activatable by adding `DBusActivatable=true` to the `.desktop` file (the `Exec=` line can be kept for backwards compatibility). 

When a DBus activatable application is launched by the desktop manager, instead of directly starting Ghostty, the desktop manager will send a message to the DBus session manager asking it to start up Ghostty. If there is no process that has claimed Ghostty's DBus name (`com.mitchellh.ghostty` by default, but configurable) DBus will look for a DBus service file that tells it how to launch Ghostty. DBus service files can tell the DBus session manager how to start Ghostty directly, or they can tell the DBus session manager to start a systemd user service.

systemd would be the service that ultimately launches Ghostty.

Running Ghostty as a DBus activatable service has the advantage of being slightly more efficient if Ghostty is already running.

Running Ghostty as a systemd user service has several advantages.

1) output from Ghostty is collected into the journal and makes it easier to find (`journalctl --user -u com.mitchellh.ghostty.service`).
2) Ghostty can be launched into the background when the user logs (`systemctl --user enable com.mitchellh.ghostty.service` or `systemctl --user disable com.mitchellh.ghostty.service`)
3) A Ghostty running in the background can be shut down (`systemctl --user stop com.mitchellh.ghostty.service`) or restarted (`systemctl --user restart com.mitchellh.ghostty.service`) for upgrades.

If the desktop launcher does not directly support DBus activation (like many tiling window managers), a new Ghostty window could be created with a command like this:
```
gdbus call --session --dest com.mitchellh.ghostty --object-path /com/mitchellh/ghostty --method org.gtk.Actions.Activate new-window [] []
```

Code changes in Zig should be fairly minor, mainly making sure that windows aren't created prematurely when being launched by DBus activation or by systemd.

**Comment by SuperDuperDeou** [comment]
- Ghostty should have a CLI option to activate the service. Said CLI option should have an implementation for other platforms, or GNU/Linux+Systemd machines will eventually leave other platform in the dust in functionality. 
  - non-Systemd GNU/Linux distros: They still have Dbus, so it would be activated by Dbus directly, without going through a Systemd unit.
  - MacOS: I'm not a MacOS expert, but it's still a BSD, it should have sockets.
  - Windows: Windows has its own service API, AFAIK.
- There should be a way to activate a service without needing a window open and for the process to not kill itself when the last window is closed, a proper daemon mode, so that the activation only needs to be done once at log time and never again.

Once all that is figured out, this can be expanded upon to implement other features, such as remote control ( #2353 ) and session management ( #3358  )

**Reply by jcollie** [reply]
You can do what you want right now: `ghostty --gtk-single-instance=true --quit-after-last-window-closed=false --initial-window=false` will start up Ghostty but not create a window and it will never quit. Then the next time you "launch" Ghostty the new process should tell the "daemon" to create a new window through a DBus call (as long as `--gtk-single-instance=true`). 

None of this is necessary on macOS as you can't have multiple copies of the application running at once (at least not without going to a lot of trouble).

When Windows support is worked on I'm sure that a "daemon" mode will be figured out.

I disagree that there should be a single "--daemon" flag as that would be difficult/confusing to implement across all platforms.

The point of this discussion is to make "daemon" mode _even better_ on Linux. DBus activation and a systemd user unit are IMHO the best way to do that.

**Reply by SuperDuperDeou** [reply]
I agree that supporting Dbus is a good thing, but Systemd isn't found on every system, and Dbus doesn't require Systemd to start a service. If anything, writing a user init script for a foo init system can be left to the distros.

The Dbus interface can be expanded upon to provide more functionality, which makes it extra interesting. But instead of Dbus, it could be just a socket, and Ghostty would have the same infrastructure in place for both GNU/Linux and MacOS, with the FreeDesktop-compliant Unixes having less to worry about when making their own port. It could really be either way.

The implementation would be just as easy: on FreeDesktop system, on startup, Ghostty would try to bind `${XDG_RUNTIME_DIR}/ghostty_in.sock` and `${XDG_RUNTIME_DIR}/ghostty_out.sock`, if they already exist it means that there is a Ghostty server already running, and it the new and old processes negotiate over the sockets.

**Reply by jcollie** [reply]
The systemd user service and dbus activation services will be optional. If your system doesn't use system, don't install them. There are some internal changes to Ghostty that need to happen to make systemd user services and dbus activation work optimally, but that will be transparent to any systems that don't use them.

**Reply by SuperDuperDeou** [reply]
IMHO, an init script should be left to the distros.

Dbus, at the end of the day, is a socket router. What we're doing it, is adding a way for a terminal to talk with other instances of itself, not to make a cross-desktop standard that should work with who-knows-what combination of software, I see no reason to not just cut out the middle man and use sockets directly.

**Reply by jcollie** [reply]
One of the goals for Ghostty is a _native_ experience. Gnome/GTK rely heavily on DBus and DBus activation. Using DBus in this way isn't _just_ a way for Ghostty to talk to itself, but a way for Ghostty to talk to Gnome in the way that Gnome expects.

A custom socket will likely get added at some point for API access to Ghostty, but then that _will_ be cross-platform as it will be expected to work identically across Linux, macOS and eventually Windows.

**Reply by SuperDuperDeou** [reply]
Native to what? The website doesn't mention native to GNOME, but native to "Linux", whatever that means since Linux isn't a desktop, isn't a desktop standard, isn't an OS, it's a kernel.

If the goal from the start was making a GNOME app and disregard non-GNOME system, the `sed s/Linux/GNOME/g` the documentation.

**Reply by tristan957** [reply]
We should keep init scripts upstream. It makes no sense for distros to duplicate work or to not ship a good experience to users because they didn't copy the patch from a different distro.

**Comment by jcollie** [comment]
One thing that I discovered that I was able to do today with this setup is to create a global shortcut in Gnome that will open a new Ghostty window, even if Ghostty is not running. Basically in the Gnome settings menu create a new custom shortcut with this as the command:

```
gdbus call --session --dest com.mitchellh.ghostty --object-path /com/mitchellh/ghostty --method org.gtk.Actions.Activate new-window [] []
```

That will use DBus activation to start up the Ghostty user service and then tells Ghostty to create a new window.

![Screenshot From 2024-12-29 01-41-30](https://github.com/user-attachments/assets/14e4bfdd-502a-4260-aa81-e95f4a0a24bb)

**Reply by gerblesh** [reply]
Was unable to run this command on sway:
```
$ gdbus call --session --dest com.mitchellh.ghostty --object-path /com/mitchellh/ghostty --method org.gtk.Actions.Activate new-window [] []
Error: GDBus.Error:org.freedesktop.DBus.Error.InvalidArgs: Unknown action ‘new-window’
(According to introspection data, you need to pass 'sava{sv}')
```

Do I need to be using GNOME for this?

**Reply by Opposite34** [reply]
I'm pretty sure something changed since then as I tried on GNOME just now and got the same thing with @gerblesh

**Reply by jcollie** [reply]
> Was unable to run this command on sway:
> 
> ```
> $ gdbus call --session --dest com.mitchellh.ghostty --object-path /com/mitchellh/ghostty --method org.gtk.Actions.Activate new-window [] []
> Error: GDBus.Error:org.freedesktop.DBus.Error.InvalidArgs: Unknown action ‘new-window’
> (According to introspection data, you need to pass 'sava{sv}')
> ```
> 
> Do I need to be using GNOME for this?

No, that DBus call won't work unless you're running a patch to Ghostty that I haven't pushed a PR for yet. You can find it [here](https://github.com/jcollie/ghostty/commit/07d7a3b7084ff26c8f3a94600e8109eaa30dabac) in my fork.

**Comment by rhodes-b** [comment, marked as answer]
@jcollie lets call this one done? tip has supported systemd for a bit
